### PR TITLE
Fix EZP-25366: relation fields aren't copied correctly

### PR DESCRIPTION
**Pull request by bdunogier**
> Fixes http://jira.ez.no/browse/EZP-25366

When copying content with relation fields, relations on the copy were not properly registered. This is fixed by applying the same logic used in createContent() and updateContent(): relations inside fields are recorded during copy, and registered using the persistence later at the end.

### TODO
- [x] Handle other versions
- [x] Check if something must be done about existing relations
- [x] Tests
- [x] Add integration test for Content Copy
- [x] Add integration test for Subtree Copy

### CODE REVIEW

**Comment by bdunogier** [comment]
Ping @joaoinacio & @andrerom

File: eZ/Publish/Core/Repository/ContentService.php [hunk 1]
```diff
@@ -1586,6 +1586,9 @@ public function copyContent(ContentInfo $contentInfo, LocationCreateStruct $dest
             );
         }
 
+        $relations = $locationIdToContentIdMapping = array();
```
Comment (crevillo) [inline eZ/Publish/Core/Repository/ContentService.php]: short notation? :)

**Comment by andrerom** [comment]
Looks good, as for `LocationService->copySubtree()` seen any ways yet on how we can tackle it there?

**Comment by bdunogier** [comment]
> Looks good, as for LocationService->copySubtree() seen any ways yet on how we can tackle it there?

Well, we need to apply the same processing to copies done from there as well.
` $relationProcessor->updateContentRelations()` needs to be called after `$contentHandler->copy()`. The biggest part of the work will be to update the tests, I'm afraid.

**Comment by joaoinacio** [comment]
May be wrong here as I didn't dig deep into the tests, but it seems this could have the same problem as a recent issue, whereby relations from old versions are taken into account (kept in object_link) when they shouldn't.

Example:
```
  v1 | eng-GB | -> rel A
  v2 | fre-FR | -> rel B
- v3 | eng-GB | -> rel C
```

Correct content relations would be B , C but not A.

**Comment by joaoinacio** [comment]
Also, I'm left wondering if this solution be helpful in the context of https://jira.ez.no/browse/EZP-24540 ?

**Comment by bdunogier** [comment]
> May be wrong here as I didn't dig deep into the tests, but it seems this could have the same problem as a recent issue, whereby relations from old versions are taken into account (kept in object_link) when they shouldn't.

This bugfix doesn't really intend to fix this, to be frank.

**Comment by andrerom** [comment]
Updated approach, review ping @bdunogier @Plopix  @crevillo

**Comment by Plopix** [comment]
tested with a command like this:
```php
        // subtree
        $repo->getLocationService()->copySubtree($source->location, $dest->location);

        // simple
        $locationStruct = $repo->getLocationService()->newLocationCreateStruct($dest->location->id);
        $repo->getContentService()->copyContent($source->content->contentInfo, $locationStruct);
```

And also tested in the original context of the explained bug in the issue.

ping @andrerom @bdunogier => Nice job guys ! It works!

**Comment by bdunogier** [comment]
Great job @andrerom.

+1

**Comment by andrerom** [comment]
> Great job @andrerom.

Thanks, but did actually miss adding test in RelationSearchMultivaluedBaseIntegrationTest, which given it was a duplicate of RelationSearchBaseIntegrationTest made me change them into 1 trait during last rebase. So hopefully even better now :)

**Comment by bdunogier** [comment]
>  RelationSearchMultivaluedBaseIntegrationTest, which given it was a duplicate of RelationSearchBaseIntegrationTest made me change them into 1 trait during last rebase

Yes, this is one part of tests where we have a big debt to pay.

**Comment by yannickroger** [comment]
+1

**Comment by bdunogier** [comment]
Shouldn't you squash the integration tests with the bugfix ? We do backport the tests as well, unless I am mistaken.

**Comment by andrerom** [comment]
I'll do that on the back port :)
